Imx8mm secondary boot image support

igoropaniuk opened the pull request:
1. Support for flashing secondary images via FSL fastboot
2. Switching to the proper U-Boot FIT depending on GPR10  PERSIST_SECONDARY_BOOT bit
3. Flipping PERSIST_SECONDARY_BOOT bit for the case of regular boot (so we can rebooted into recovery path by watchdog)

igoropaniuk committed:
[FIO internal] ARM: imx8m: add support for reading out the primary/secondary bmode

Add new getprisec() function, which returns 0 if the system booted
from primary copy or to 1 if the system booted from secondary copy,
by reading PERSIST_SECONDARY_BOOT bit from GPR10 register.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

igoropaniuk committed:
[FIO internal] imx8mm_evk: runtime detection raw sector offsets

Add runtime detection for the next boot image (U-Boot proper FIT) SD/MMC
sector offset.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

igoropaniuk committed:
[FIO internal] fastboot: add support for redundant boot images

Allow fastboot to flash secondary bootloader images and secondary image
table to support redundant secondary boot feature.
For additional details check SRM 6.5.4.5 Redundant boot support for
expansion device.

Secondary Image Table contains the sector of secondary bootloader image,
exluding the offset to that image (explained below in the note).

To generate the Secondary Image Table, use e.g.:
$ printf '\x0\x0\x0\x0\x0\x0\x0\x0\x33\x22\x11\x00\x00\x01\x0\x0\x00\x0\x0\x0'
  > /tmp/sit.bin
$ hexdump  -vC /tmp/sit.bin
  00000000  00 00 00 00
  00000004  00 00 00 00
  00000008  33 22 11 00 <--- This is the "tag"
  0000000c  00 10 00 00 <--- This is the "firstSectorNumber"
  00000010  00 00 00 00

You can also use NXP script from [1] for SIT generation.
Note that the firstSectorNumber is NOT the offset of the IVT, but an
offset of the IVT decremented by Image Vector Table offset (Table 6-25.
Image Vector Table Offset and Initial Load Region Size for iMX8MM),
so for secondary SPL copy at offset 0x1042 sectors,
firstSectorNumber must be 0x1000 (0x42 sectors * 512 = 0x8400 bytes offset).

uuu fastboot additional partition flashing:
FB: flash bootloader_s ../imx-boot-imx8mmevk.signed
FB: flash bootloader2_s ../u-boot-imx8mmevk.itb
FB: flash sit ../sit.bin

In order to test redundant boot board should be closed and
SD/MMC manufacture mode disabled, as secondary boot is not supported in
the SD/MMC manufacture mode by blowing DISABLE_SDMMC_MFG:
> fuse prog -y 2 1 0x00800000

[1] https://source.codeaurora.org/external/imx/imx-mkimage/tree/scripts/gen_sit.sh?h=imx_5.4.70_2.3.0
Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

igoropaniuk commented on board/freescale/imx8mm_evk/spl.c:
```diff
@@ -2,7 +2,7 @@
  * Copyright 2019 NXP
  */
-
+#define DEBUG 1
```
This from the last commit, which I added just for testing (will be dropped in non-RFC PR)

igoropaniuk commented on board/freescale/imx8mm_evk/spl.c:
```diff
@@ -282,3 +282,27 @@ int do_reset(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
+	unsigned long offset;
+
+	if (boot_secondary)
+		offset = 0x1300;
```
I've hardcoding now offsets, will switch to according Kconfig symbols.

igoropaniuk commented on drivers/fastboot/fb_fsl/fb_fsl_partitions.c:
```diff
@@ -250,15 +262,54 @@ static int _fastboot_parts_load_from_ptable(void)
+	strcpy(ptable[PTN_BOOTLOADER2_SECONDARY_INDEX].name, FASTBOOT_PARTITION_BOOTLOADER2_SECONDARY);
+	ptable[PTN_BOOTLOADER2_SECONDARY_INDEX].start =
+				(CONFIG_FSL_FASTBOOT_BOOTLOADER2_SECONDARY_OFFSET * dev_desc->blksz) / dev_desc->blksz;
+	ptable[PTN_BOOTLOADER2_SECONDARY_INDEX].length = 2100;
```
will drop that magic length numbers

igoropaniuk committed:
[FIO internal] ARM: imx8m: change primary/secondary boot modes

Adjust PERSIST_SECONDARY_BOOT bit from GPR10 register so BootROM can
boot into the right boot path: primary or secondary

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

igoropaniuk committed:
[FIO internal] imx8mm_evk: detect needed offset for next boot image

1. Dynamic detection of offset for the next boot image based on the value
SRC_GPR10_PERSIST_SECONDARY_BOOT bit.
2. In case of regular boot set that bit so we can automatically boot
into recovery boot path after WDT reset occurs.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

mike-scott commented on drivers/fastboot/fb_fsl/fb_fsl_partitions.c:
```diff
@@ -46,14 +46,27 @@ enum {
+#endif
 #ifdef CONFIG_FSL_FASTBOOT_BOOTLOADER2
 	PTN_BOOTLOADER2_INDEX,
+	PTN_SIT_INDEX,
```
Looks like `PTN_SIT_INDEX` won't get used unless `CONFIG_FSL_FASTBOOT_BOOTLOADER_SECONDARY` is defined.  Maybe move this under the #ifdef?

igoropaniuk replied:
right, will fix

mike-scott commented on board/freescale/imx8mm_evk/spl.c:
```diff
@@ -282,3 +282,32 @@ int do_reset(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
 }
+
+#ifdef CONFIG_SPL_MMC_SUPPORT && SECONDARY_BOOT_RUNTIME_DETECTION
+unsigned long spl_mmc_get_uboot_raw_sector(struct mmc *mmc)
```
This doesn't have the same function signature as here (where it is called):
https://github.com/foundriesio/u-boot/blob/2020.04%2Bfio/common/spl/spl_mmc.c#L415

```
		raw_sect = spl_mmc_get_uboot_raw_sector(mmc, raw_sect);
```

igoropaniuk replied:
just borrowed that from another board file (`imx8mn_evk`):
```
u-boot-fio$ grep -e spl_mmc_get_uboot_raw_sector -r ./
./arch/arm/mach-sunxi/board.c:unsigned long spl_mmc_get_uboot_raw_sector(struct mmc *mmc)
./arch/arm/mach-imx/imx8/image.c:unsigned long spl_mmc_get_uboot_raw_sector(struct mmc *mmc,
./common/spl/spl_mmc.c:unsigned long __weak spl_mmc_get_uboot_raw_sector(struct mmc *mmc,
./common/spl/spl_mmc.c:		raw_sect = spl_mmc_get_uboot_raw_sector(mmc, raw_sect);
./board/freescale/imx8mn_evk/spl.c:unsigned long spl_mmc_get_uboot_raw_sector(struct mmc *mmc)
./board/freescale/imx8mp_evk/imx8mp_evk.c:unsigned long spl_mmc_get_uboot_raw_sector(struct mmc *mmc)
./board/toradex/verdin-imx8mp/verdin-imx8mp.c:unsigned long spl_mmc_get_uboot_raw_sector(struct mmc *mmc)
...
```
Interesting that compiler doesn't catch that and mark as error.

mike-scott reviewed (changes requested):
Couple of minor comments

ricardosalveti commented on arch/arm/mach-imx/imx8m/Kconfig:
```diff
@@ -27,6 +27,14 @@ config IMX8MP
 	default "imx8m"
 
+config SECONDARY_BOOT_RUNTIME_DETECTION
+	bool "SD/MMC sector offset runtine detection"
```
s/runtine/runtime/g

ricardosalveti commented:
https://source.codeaurora.org/external/imx/imx-mkimage/tree/scripts/gen_sit.sh (master) gives a not found. Probably wanted to link to a more recent BSP branch?

ricardosalveti commented on drivers/fastboot/fb_fsl/fb_fsl_partitions.c:
```diff
@@ -46,14 +46,27 @@ enum {
 struct fastboot_ptentry g_ptable[MAX_PTN];
 unsigned int g_pcount;
 
+static ulong secondary_image_table_mmc_offset(void)
```
Might cause a defined but not used warning when CONFIG_FSL_FASTBOOT_BOOTLOADER_SECONDARY is not defined.

ricardosalveti reviewed (changes requested):
Just a few minor comments as well.

igoropaniuk commented:
@ricardosalveti @mike-scott all issues addressed

ricardosalveti commented on board/freescale/imx8mm_evk/spl.c:
```diff
@@ -282,3 +282,33 @@ int do_reset(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
+		/*
+		 * Set GPR10 PERSIST_SECONDARY_BOOT bit,
+		 * so in case of hang or other issues
+		 * Watchdog will us into recovery boot path
```
watchdog will move?

igoropaniuk replied:
`boot us`

igoropaniuk replied:
fixed

ricardosalveti reviewed (approved):
LGTM, just one minor comment.

mike-scott commented on drivers/fastboot/fb_fsl/fb_fsl_partitions.c:
```diff
@@ -250,15 +265,59 @@ static int _fastboot_parts_load_from_ptable(void)
-	ptable[PTN_BOOTLOADER2_INDEX].length =
-				(ANDROID_BOOTLOADER_SIZE - (CONFIG_FSL_FASTBOOT_BOOTLOADER2_OFFSET * dev_desc->blksz))
-				/ dev_desc->blksz;
+	ptable[PTN_BOOTLOADER2_INDEX].length = (ANDROID_BOOTLOADER_SIZE -
```
This is now a formatting change.  Can it be dropped?

mike-scott commented on drivers/fastboot/fb_fsl/fb_fsl_partitions.c:
```diff
@@ -239,7 +254,7 @@ static int _fastboot_parts_load_from_ptable(void)
 #else
 	ptable[PTN_BOOTLOADER_INDEX].length =
-				 ANDROID_BOOTLOADER_SIZE / dev_desc->blksz;
+				 (ANDROID_BOOTLOADER_SIZE - bootloader_mmc_offset()) / dev_desc->blksz;
```
In a way, this is a bugfix change and not related to this patch.   Maybe split out?

mike-scott commented on drivers/fastboot/fb_fsl/fb_fsl_partitions.c:
```diff
@@ -250,15 +265,59 @@ static int _fastboot_parts_load_from_ptable(void)
+	 * Bootloader images for Secondary rendundant boot
+	 */
+	strcpy(ptable[PTN_BOOTLOADER_SECONDARY_INDEX].name, FASTBOOT_PARTITION_BOOTLOADER_SECONDARY);
+	ptable[PTN_BOOTLOADER_SECONDARY_INDEX].start = CONFIG_SECONDARY_BOOT_SECTOR_OFFSET +
```
In theory this is the same as:
```
ptable[PTN_BOOTLOADER_SECONDARY_INDEX].start = CONFIG_SECONDARY_BOOT_SECTOR_OFFSET + ptable[PTN_BOOTLOADER_INDEX].start;
```

mike-scott commented on drivers/fastboot/fb_fsl/fb_fsl_partitions.c:
```diff
@@ -250,15 +265,59 @@ static int _fastboot_parts_load_from_ptable(void)
+	strcpy(ptable[PTN_BOOTLOADER_SECONDARY_INDEX].name, FASTBOOT_PARTITION_BOOTLOADER_SECONDARY);
+	ptable[PTN_BOOTLOADER_SECONDARY_INDEX].start = CONFIG_SECONDARY_BOOT_SECTOR_OFFSET +
+		(bootloader_mmc_offset() / dev_desc->blksz);
+#ifdef CONFIG_FSL_FASTBOOT_BOOTLOADER2
```
This #if #else block can be dropped with:
```
ptable[PTN_BOOTLOADER_SECONDARY_INDEX].length = ptable[PTN_BOOTLOADER_INDEX].length;
```

mike-scott commented on drivers/fastboot/fb_fsl/fb_fsl_partitions.c:
```diff
@@ -250,15 +265,59 @@ static int _fastboot_parts_load_from_ptable(void)
+#ifdef CONFIG_FSL_FASTBOOT_BOOTLOADER2
+	/* Bootloader2 (u-boot binary when SPL is separate) */
+	strcpy(ptable[PTN_BOOTLOADER2_SECONDARY_INDEX].name, FASTBOOT_PARTITION_BOOTLOADER2_SECONDARY);
+	ptable[PTN_BOOTLOADER2_SECONDARY_INDEX].start =
```
Same as:
```
ptable[PTN_BOOTLOADER2_SECONDARY_INDEX].start = CONFIG_SECONDARY_BOOT_SECTOR_OFFSET + ptable[PTN_BOOTLOADER2_INDEX].start;
```

mike-scott commented on drivers/fastboot/fb_fsl/fb_fsl_partitions.c:
```diff
@@ -250,15 +265,59 @@ static int _fastboot_parts_load_from_ptable(void)
+	ptable[PTN_BOOTLOADER2_SECONDARY_INDEX].start =
+				((CONFIG_SECONDARY_BOOT_SECTOR_OFFSET + CONFIG_FSL_FASTBOOT_BOOTLOADER2_OFFSET) *
+				 dev_desc->blksz) / dev_desc->blksz;
+	ptable[PTN_BOOTLOADER2_SECONDARY_INDEX].length = ((ANDROID_BOOTLOADER_SIZE * 2) -
```
Same as:
```
ptable[PTN_BOOTLOADER2_SECONDARY_INDEX].length = ptable[PTN_BOOTLOADER2_INDEX].length;
```

mike-scott commented on arch/arm/mach-imx/imx8m/Kconfig:
```diff
@@ -27,6 +27,14 @@ config IMX8MP
+config SECONDARY_BOOT_RUNTIME_DETECTION
+	bool "SD/MMC sector offset runtime detection"
+	default n
+	depends on (IMX8MQ || IMX8MM) && SYS_MMCSD_RAW_MODE_U_BOOT_USE_SECTOR
```
Should we depend on SPL_MMC_SUPPORT as well here?

ricardosalveti replied:
Wonder if needed because of SYS_MMCSD_RAW_MODE_U_BOOT_USE_SECTOR, but looking at Kconfig this is not checking for MMC either, so not sure.

mike-scott commented on board/freescale/imx8mm_evk/spl.c:
```diff
@@ -282,3 +282,34 @@ int do_reset(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
 	return 0;
 }
+
+#ifdef CONFIG_SPL_MMC_SUPPORT && SECONDARY_BOOT_RUNTIME_DETECTION
```
This is broken currently and luckily evaluates to true because `CONFIG_SPL_MMC_SUPPORT` is defined.  To work correctly this would need to be:
```
#if defined(CONFIG_SPL_MMC_SUPPORT) && defined(CONFIG_SECONDARY_BOOT_RUNTIME_DETECTION)
```

And if we add `CONFIG_SPL_MMC_SUPPORT` to the Kconfig depends we can drop it here:
```
#ifdef CONFIG_SECONDARY_BOOT_RUNTIME_DETECTION
```

igoropaniuk replied:
yeah, right

mike-scott commented on include/configs/imx8mm_evk.h:
```diff
@@ -12,8 +12,10 @@
 #define CONFIG_SYS_MONITOR_LEN		SZ_512K
+#ifndef CONFIG_SECONDARY_BOOT_RUNTIME_DETECTION
 #define CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_USE_SECTOR
 #define CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_SECTOR	(0x300 + CONFIG_SECONDARY_BOOT_SECTOR_OFFSET)
```
Not sure why we are setting this to 0x300 + CONFIG_SECONDARY_BOOT_SECTOR_OFFSET when this is #ifndef CONFIG_SECONDARY_BOOT_RUNTIME_DETECTION

mike-scott replied:
Ah.. I see.  Original intent of `CONFIG_SECONDARY_BOOT_SECTOR_OFFSET` was to be a compile time setting for u-boot.  We want to avoid this case when we are using RUNTIME detection.

Ignore this.

mike-scott replied:
Or maybe add a comment above the #ifndef to explain the intention.

mike-scott replied:
Actually, I would like to see this here (leaving all settings as defined but with correct values):
```
#define CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_USE_SECTOR
/*
 * CONFIG_SECONDARY_BOOT_SECTOR_OFFSET was intended to be used as a compile time setting.
 * If CONFIG_SECONDARY_BOOT_RUNTIME_DETECTION is defined, then only use the base offset of 0x300 here.
 * CONFIG_SECONDARY_BOOT_SECTOR_OFFSET will be added dynamically if needed.
 */
#ifdef CONFIG_SECONDARY_BOOT_RUNTIME_DETECTION
#define CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_SECTOR	(0x300)
#else
#define CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_SECTOR	(0x300 + CONFIG_SECONDARY_BOOT_SECTOR_OFFSET)
#endif /* CONFIG_SECONDARY_BOOT_RUNTIME_DETECTION */
```

igoropaniuk replied:
I think defining `CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_SECTOR` here instead of defconfig is a bad idea from the very beginning. The reason I wrapped that in that way was to avoid build issues (if we decide to build `imx8mm_evk_defconfig` outside our lmp OE setup)

mike-scott commented on include/configs/imx8mm_evk.h:
```diff
@@ -12,8 +12,10 @@
+#ifndef CONFIG_SECONDARY_BOOT_RUNTIME_DETECTION
 #define CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_USE_SECTOR
 #define CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_SECTOR	(0x300 + CONFIG_SECONDARY_BOOT_SECTOR_OFFSET)
+#endif /* CONFIG_SECONDARY_BOOT_RUNTIME_DETECTION */
```
```
#endif /* !CONFIG_SECONDARY_BOOT_RUNTIME_DETECTION */
```

igoropaniuk committed:
[FIO internal] fastboot: fix length calculation for imx-boot image

Fix length calculation for imx-boot image, take into account additional
offset to IVT.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

mike-scott commented on board/freescale/imx8mm_evk/spl.c:
```diff
@@ -282,3 +282,34 @@ int do_reset(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
 	return 0;
 }
+
+#ifdef SECONDARY_BOOT_RUNTIME_DETECTION
```
#ifdef CONFIG_SECONDARY_BOOT_RUNTIME_DETECTION

igoropaniuk replied:
good catch, thanks!
fixed

mike-scott commented on arch/arm/mach-imx/imx8m/soc.c:
```diff
@@ -608,6 +608,29 @@ enum boot_device get_boot_device(void)
+	setbits_le32(&psrc->gpr10, persist_sec);
+}
+
+void boot_mode_set_primary(void)
```
@igoropaniuk Is this used?

mike-scott commented on arch/arm/mach-imx/imx8m/soc.c:
```diff
@@ -608,6 +608,29 @@ enum boot_device get_boot_device(void)
+	return !!(readl(&psrc->gpr10) & SRC_GPR10_PERSIST_SECONDARY_BOOT);
+}
+
+void boot_mode_set_secondary(void)
```
@ricardosalveti @igoropaniuk how would we feel about combining  this into:
```
void boot_mode_enable_secondary(unsigned char enable)
{
	u32 persist_sec = SRC_GPR10_PERSIST_SECONDARY_BOOT;
	struct src *psrc = (struct src *)SRC_BASE_ADDR;

	if (enable)
		setbits_le32(&psrc->gpr10, persist_sec);
	else
		clrbits_le32(&psrc->gpr10, persist_sec);
}
```

igoropaniuk replied:
I've added the function you suggested, with only change
s/unsigned char enable/bool/g

mike-scott commented on board/freescale/imx8mm_evk/spl.c:
```diff
@@ -282,3 +282,34 @@ int do_reset(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
+		 * BootROM will load recovery boot images
+		 * after _warm_ reset.
+		 */
+		boot_mode_set_secondary();
```
if we combined the 2 functions this would be:
```
boot_mode_enable_secondary(1);
```

mike-scott reviewed (commented):
Looks good just the question on combining the secondary mode function

mike-scott commented on arch/arm/include/asm/arch-imx8m/sys_proto.h:
```diff
@@ -14,5 +14,9 @@ void restore_boot_params(void);
 extern unsigned long rom_pointer[];
 enum boot_device get_boot_device(void);
+int boot_mode_getprisec(void);
+void boot_mode_set_primary(void);
```
These 2 can be dropped.

mike-scott reviewed (approved):
LGTM

ricardosalveti commented on board/freescale/imx8mm_evk/spl.c:
```diff
@@ -282,3 +282,34 @@ int do_reset(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
+unsigned long spl_mmc_get_uboot_raw_sector(struct mmc *mmc,
+					   unsigned long raw_sect)
+{
+
```
Minor, no need for this extra empty line.
